repository-commons/src/main/java/io/aiven/elasticsearch/repositories/ClientProvider.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2021 Aiven Oy
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.elasticsearch.repositories;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Objects;
+
+import org.opensearch.common.settings.Settings;
+
+public abstract class ClientProvider<C, S extends CommonSettings.ClientSettings> implements Closeable {
+
+    private final Object lock = new Object();
+
+    private volatile Settings previousRepositorySettings;
+
+    protected volatile C client;
+
+    public C buildClientIfNeeded(final S clientSettings, final Settings repositorySettings) throws IOException {
+        if (Objects.isNull(client)) {
+            client = buildClient0(clientSettings, repositorySettings);
+        } else if (!previousRepositorySettings.equals(repositorySettings)) {
+            close();
+            client = buildClient0(clientSettings, repositorySettings);
+        }
+        return client;
+    }
+
+    private C buildClient0(final S clientSettings, final Settings repositorySettings) {
+        synchronized (lock) {
+            client = buildClient(clientSettings, repositorySettings);
```

Comment:
You need to check `if (Objects.isNull(client)) {`, it could happen than more than one thread will arrive there,https://github.com/aiven/aiven-elasticsearch-repositories/pull/62/files#diff-466040efd623e1f83d78c6c2545fd1fbeca265f85d1cb62d33d8106a711a026eR35, each of them will get the lock and create client, one after another.

Reply:
True. Will fix